scale-tests/kafka_cassandra_streaming_test.py

```diff
@@ -175,58 +202,70 @@ def append_submission(output_file: str, dispatcher_app_name: str, submission_id:
-        zookeeper = infrastructure['zookeeper'][0]
-        kafka = infrastructure['kafka'][0]
+        zookeepers = infrastructure['zookeeper']
+        kafkas = infrastructure['kafka']
         cassandra = infrastructure['cassandra'][0]
 
     jar                           = args["--jar"] if args["--jar"] else DEFAULT_JAR
     submissions_output_file       = args["<submissions_output_file>"]
-    kafka_package_name            = kafka['package_name']
-    kafka_service_name            = kafka['service']['name']
-    kafka_num_brokers             = kafka['brokers']['count']
+    kafka_package_name            = kafkas[0]['package_name']
     cassandra_package_name        = cassandra['package_name']
     cassandra_service_name        = cassandra['service']['name']
     cassandra_num_nodes           = cassandra['nodes']['count']
-    num_producers_per_dispatcher  = int(args['--num-producers-per-dispatcher'])
+    num_producers_per_kafka       = int(args['--num-producers-per-kafka'])
     num_consumers_per_producer    = int(args['--num-consumers-per-producer'])
+    producer_must_fail            = args['--producer-must-fail']
     producer_number_of_words      = int(args['--producer-number-of-words'])
     producer_words_per_second     = int(args['--producer-words-per-second'])
     producer_spark_cores_max      = int(args['--producer-spark-cores-max'])
     producer_spark_executor_cores = int(args['--producer-spark-executor-cores'])
+    consumer_must_fail            = args['--consumer-must-fail']
     consumer_write_to_cassandra   = args['--consumer-write-to-cassandra']
     consumer_batch_size_seconds   = int(args['--consumer-batch-size-seconds'])
     consumer_spark_cores_max      = int(args['--consumer-spark-cores-max'])
     consumer_spark_executor_cores = int(args['--consumer-spark-executor-cores'])
 
-    producer_must_fail = args['--producer-must-fail']
-    consumer_must_fail = args['--consumer-must-fail']
-
-    log.info("Dispatchers: \n{}".format("\n".join(dispatchers)))
+    num_kafkas = len(kafkas)
+    num_dispatchers = len(dispatchers)
+    num_producers = num_kafkas * num_producers_per_kafka
+    num_consumers = num_producers * num_consumers_per_producer
+    num_jobs = num_producers + num_consumers
+    avg_num_jobs_per_dispatcher = num_jobs / num_dispatchers
+    max_num_jobs_per_dispatcher = math.ceil(avg_num_jobs_per_dispatcher)
+
+    log.info('Number of Kafka clusters: {}'.format(num_kafkas))
+    log.info('Number of dispatchers: {}'.format(num_dispatchers))
+    log.info('Total number of jobs: {} ({} producers, {} consumers)'.format(
+        num_jobs, num_producers, num_consumers))
+    log.info('Average number jobs per dispatcher: {}'.format(avg_num_jobs_per_dispatcher))
+    log.info('Will run at most {} jobs per dispatcher'.format(max_num_jobs_per_dispatcher))
+    log.info("\nDispatchers: \n{}\n".format("\n".join(dispatchers)))
+
+    dispatchers_slots = sdk_utils.mapcat(make_repeater(max_num_jobs_per_dispatcher),
```

Comment:
This means that we are allocating to dispatchers in blocks? Would round-robin be better?

Reply:
When first reading this line, it also was not clear what we were doing here. We could move it to a function to make the intent here clearer?

Reply:
Yes. This line generates a list of repeated dispatcher entries.

```python
>>> dispatchers = ['dispatcher-0', 'dispatcher-1']
>>> sdk_utils.mapcat(make_repeater(3), dispatchers)
['dispatcher-0', 'dispatcher-0', 'dispatcher-0', 'dispatcher-1', 'dispatcher-1', 'dispatcher-1']
```
I'll try to make it clearer...